apiserver/apiserver/coordinator/matchmaking.py

```diff
@@ -129,9 +135,114 @@ def serve_game_task(conn, has_gpu=False):
             "width": map_width,
             "height": map_height,
             "users": players,
+            "challenge": None,
         })
 
 
+def reset_challenges(conn):
+    """Check ongoing challenges, and reset ones that are "stuck"."""
+    reset_stuck_challenges = model.challenges.update().where(
+        (model.challenges.c.status == model.ChallengeStatus.PLAYING_GAME.value) &
+        (model.challenges.c.most_recent_game_task <
+         datetime.datetime.now() - datetime.timedelta(
+             minutes=30))
+    ).values(
+        status=model.ChallengeStatus.CREATED.value,
+    )
+    conn.execute(reset_stuck_challenges)
+
+
+def find_challenge(conn, has_gpu=False):
+    """
+    Find a set of players in a challenge.
+    """
+    # Reset any stuck challenges
+    reset_challenges(conn)
+
+    challenge = conn.execute(
+        model.challenges.select(
+            model.challenges.c.status == model.ChallengeStatus.CREATED.value
```

Comment:
This query should incorporate getting 'stuck' challenges, something like:

    challenge = conn.execute(
        model.challenges.select(
            model.challenges.c.status == model.ChallengeStatus.CREATED.value
            | (model.challenges.c.status == model.ChallengeStatus.PLAYING_GAME.value
               & model.challenges.c.most_recent_game_task <
                 datetime.datetime.now() - datetime.timedelta(minutes=30))
        ).order_by(
            model.challenges.c.most_recent_game_task.asc()
        )
    ).first()

Reply:
The `reset_challenges` function defined above resets any such challenges, though integrating them into the query also makes sense.

Reply:
Yep and I think there is value in keeping the `reset_challenges` button. This just allows them to get picked up even if the full reset isn't done.